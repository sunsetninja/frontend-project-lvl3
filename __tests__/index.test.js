/* eslint-disable functional/no-let */
/* eslint-disable no-underscore-dangle */
import { beforeEach, test } from '@jest/globals';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import nock from 'nock';
import testingLibrary from '@testing-library/dom';
import testingLibraryUserEvent from '@testing-library/user-event';
import '@testing-library/jest-dom/extend-expect';
import runApp from '../src/application.js';

const userEvent = testingLibraryUserEvent.default;

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const { screen, waitFor } = testingLibrary;

const rssUrl1 = 'https://lorem.com/feed1.rss';
let rssData1;
const rssUrl2 = 'https://lorem.com/feed1.rss';
let rssData2;

const getFixture = (name) => fs.readFileSync(path.resolve(__dirname, '__fixtures__', name));

beforeAll(() => {
  rssData1 = getFixture('response1.rss').toString();
  rssData2 = getFixture('response2.rss').toString();
  nock.disableNetConnect();

  nock('https://hexlet-allorigins.herokuapp.com')
    .persist()
    .defaultReplyHeaders({
      'access-control-allow-origin': '*',
      'access-control-allow-credentials': 'true',
    })
    .get('/get')
    .query({ url: rssUrl1, disableCache: 'true' })
    .reply(200, { contents: rssData1 })
    .get('/get')
    .query({ url: rssUrl2, disableCache: 'true' })
    .reply(200, { contents: rssData2 });
});

const domparser = new DOMParser();

beforeEach(() => {
  const initDOM = domparser.parseFromString(
    fs.readFileSync(path.resolve(__dirname, '../index.html')).toString(),
    'text/html',
  );
  document.body.innerHTML = initDOM.querySelector('body').innerHTML;
  runApp();
});

test('initial app state', async () => {
  const urlEl = screen.getByRole('textbox', { name: /url/ });
  const feedbackEl = screen.getByTestId('form-feedback');
  const feedsEl = screen.getByTestId('feeds');
  const postsEl = screen.getByTestId('posts');

  expect(urlEl).toHaveValue('');
  expect(feedbackEl).toHaveTextContent('');
  expect(feedsEl).toHaveTextContent('');
  expect(postsEl).toHaveTextContent('');
});

test('invalid rss url', async () => {
  const submitEl = screen.getByRole('button', { name: /add/ });
  const urlEl = screen.getByRole('textbox', { name: /url/ });
  const feedbackEl = screen.getByTestId('form-feedback');

  await userEvent.type(urlEl, 'invalid_url');
  await userEvent.click(submitEl);

  expect(urlEl).toHaveClass('is-invalid');
  expect(feedbackEl).toHaveTextContent(/must be valid url/i);
});

test('existed rss url', async () => {
  const submitEl = screen.getByRole('button', { name: /add/ });
  const urlEl = screen.getByRole('textbox', { name: /url/ });
  const feedbackEl = screen.getByTestId('form-feedback');

  await userEvent.type(urlEl, rssUrl1.toString());
  await userEvent.click(submitEl);

  await waitFor(async () => {
    screen.getByText(/rss has been loaded/i);
  });

  await userEvent.type(urlEl, rssUrl1.toString());
  await userEvent.click(submitEl);

  expect(urlEl).toHaveClass('is-invalid');
  expect(feedbackEl).toHaveTextContent(/rss already exists/i);
});

test('valid rss urls', async () => {
  const submitEl = screen.getByRole('button', { name: /add/ });
  const urlEl = screen.getByRole('textbox', { name: /url/ });
  const feedbackEl = screen.getByTestId('form-feedback');
  const feedsEl = screen.getByTestId('feeds');
  const postsEl = screen.getByTestId('posts');

  await userEvent.type(urlEl, rssUrl1.toString());
  await userEvent.click(submitEl);

  expect(urlEl).toHaveAttribute('readonly');
  expect(submitEl).toBeDisabled();

  await waitFor(() => {
    expect(feedbackEl).toHaveTextContent(/rss has been loaded/i);
  });

  await userEvent.type(urlEl, rssUrl2.toString());
  await userEvent.click(submitEl);

  await waitFor(() => {
    expect(feedsEl).toMatchSnapshot();
    expect(postsEl).toMatchSnapshot();
  });
});

test('post preview open', async () => {
  const submitEl = screen.getByRole('button', { name: /add/ });
  const urlEl = screen.getByRole('textbox', { name: /url/ });
  const postsEl = screen.getByTestId('posts');
  const modalEl = screen.getByTestId('modal');

  await userEvent.type(urlEl, rssUrl1.toString());
  await userEvent.click(submitEl);

  await waitFor(async () => {
    screen.getByText(/rss has been loaded/i);
  });

  await userEvent.click(screen.getAllByRole('button', { name: /preview/i })[0]);

  await waitFor(() => {
    expect(modalEl).toHaveClass('show');
    expect(modalEl).toMatchSnapshot();
    expect(postsEl).toMatchSnapshot();
  });
});
