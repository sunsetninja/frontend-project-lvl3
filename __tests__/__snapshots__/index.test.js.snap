// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`post preview open 1`] = `
<div
  aria-labelledby="modalLabel"
  aria-modal="true"
  class="modal fade show"
  data-component="modal"
  data-testid="modal"
  id="modal"
  role="dialog"
  style="padding-right: 0px; display: block;"
  tabindex="-1"
>
  
      
  <div
    class="modal-dialog"
  >
    
        
    <div
      class="modal-content"
    >
      
          
      <div
        class="modal-header"
      >
        
            
        <h5
          class="modal-title"
          data-element="modal-title"
        >
          Lorem ipsum 2021-01-15T08:33:00Z
        </h5>
        
            
        <button
          aria-label="Close"
          class="btn-close"
          data-bs-dismiss="modal"
          type="button"
        />
        
          
      </div>
      
          
      <div
        class="modal-body"
        data-element="modal-body"
      >
        Laborum nulla cupidatat labore aliquip minim enim mollit sint nisi tempor aliquip nisi cupidatat.
      </div>
      
          
      <div
        class="modal-footer"
      >
        
            
        <a
          class="btn btn-primary"
          data-element="modal-full-article-link"
          href="http://example.com/test/1610699580"
          rel="noopener noreferrer"
          role="button"
          target="_blank"
        >
          Full article
        </a>
        
            
        <button
          class="btn btn-secondary"
          data-bs-dismiss="modal"
          type="button"
        >
          
              Close
            
        </button>
        
          
      </div>
      
        
    </div>
    
      
  </div>
  
    
</div>
`;

exports[`post preview open 2`] = `
<div
  class="col-md-10 col-lg-8 mx-auto posts"
  data-component="posts"
  data-testid="posts"
>
  
    
  <h2>
    Posts
  </h2>
  
    
  <ul
    class="list-group"
  >
    
      

    <li
      class="list-group-item d-flex justify-content-between align-items-start"
    >
      
  
      <a
        class="fw-normal font-weight-normal"
        data-id="2"
        href="http://example.com/test/1610699580"
        rel="noopener noreferrer"
        target="_blank"
      >
        
    Lorem ipsum 2021-01-15T08:33:00Z
  
      </a>
      
  
      <button
        class="btn btn-primary btn-sm"
        data-bs-target="#modal"
        data-bs-toggle="modal"
        data-id="http://example.com/test/1610699580"
        type="button"
      >
        
    Preview
  
      </button>
      

    </li>
    


    <li
      class="list-group-item d-flex justify-content-between align-items-start"
    >
      
  
      <a
        class="fw-bold font-weight-bold"
        data-id="2"
        href="http://example.com/test/1610699520"
        rel="noopener noreferrer"
        target="_blank"
      >
        
    Lorem ipsum 2021-01-15T08:32:00Z
  
      </a>
      
  
      <button
        class="btn btn-primary btn-sm"
        data-bs-target="#modal"
        data-bs-toggle="modal"
        data-id="http://example.com/test/1610699520"
        type="button"
      >
        
    Preview
  
      </button>
      

    </li>
    


    <li
      class="list-group-item d-flex justify-content-between align-items-start"
    >
      
  
      <a
        class="fw-bold font-weight-bold"
        data-id="2"
        href="http://example.com/test/1610699460"
        rel="noopener noreferrer"
        target="_blank"
      >
        
    Lorem ipsum 2021-01-15T08:31:00Z
  
      </a>
      
  
      <button
        class="btn btn-primary btn-sm"
        data-bs-target="#modal"
        data-bs-toggle="modal"
        data-id="http://example.com/test/1610699460"
        type="button"
      >
        
    Preview
  
      </button>
      

    </li>
    


    <li
      class="list-group-item d-flex justify-content-between align-items-start"
    >
      
  
      <a
        class="fw-bold font-weight-bold"
        data-id="2"
        href="http://example.com/test/1610699400"
        rel="noopener noreferrer"
        target="_blank"
      >
        
    Lorem ipsum 2021-01-15T08:30:00Z
  
      </a>
      
  
      <button
        class="btn btn-primary btn-sm"
        data-bs-target="#modal"
        data-bs-toggle="modal"
        data-id="http://example.com/test/1610699400"
        type="button"
      >
        
    Preview
  
      </button>
      

    </li>
    


    <li
      class="list-group-item d-flex justify-content-between align-items-start"
    >
      
  
      <a
        class="fw-bold font-weight-bold"
        data-id="2"
        href="http://example.com/test/1610699340"
        rel="noopener noreferrer"
        target="_blank"
      >
        
    Lorem ipsum 2021-01-15T08:29:00Z
  
      </a>
      
  
      <button
        class="btn btn-primary btn-sm"
        data-bs-target="#modal"
        data-bs-toggle="modal"
        data-id="http://example.com/test/1610699340"
        type="button"
      >
        
    Preview
  
      </button>
      

    </li>
    


    <li
      class="list-group-item d-flex justify-content-between align-items-start"
    >
      
  
      <a
        class="fw-bold font-weight-bold"
        data-id="2"
        href="http://example.com/test/1610699280"
        rel="noopener noreferrer"
        target="_blank"
      >
        
    Lorem ipsum 2021-01-15T08:28:00Z
  
      </a>
      
  
      <button
        class="btn btn-primary btn-sm"
        data-bs-target="#modal"
        data-bs-toggle="modal"
        data-id="http://example.com/test/1610699280"
        type="button"
      >
        
    Preview
  
      </button>
      

    </li>
    


    <li
      class="list-group-item d-flex justify-content-between align-items-start"
    >
      
  
      <a
        class="fw-bold font-weight-bold"
        data-id="2"
        href="http://example.com/test/1610699220"
        rel="noopener noreferrer"
        target="_blank"
      >
        
    Lorem ipsum 2021-01-15T08:27:00Z
  
      </a>
      
  
      <button
        class="btn btn-primary btn-sm"
        data-bs-target="#modal"
        data-bs-toggle="modal"
        data-id="http://example.com/test/1610699220"
        type="button"
      >
        
    Preview
  
      </button>
      

    </li>
    


    <li
      class="list-group-item d-flex justify-content-between align-items-start"
    >
      
  
      <a
        class="fw-bold font-weight-bold"
        data-id="2"
        href="http://example.com/test/1610699160"
        rel="noopener noreferrer"
        target="_blank"
      >
        
    Lorem ipsum 2021-01-15T08:26:00Z
  
      </a>
      
  
      <button
        class="btn btn-primary btn-sm"
        data-bs-target="#modal"
        data-bs-toggle="modal"
        data-id="http://example.com/test/1610699160"
        type="button"
      >
        
    Preview
  
      </button>
      

    </li>
    


    <li
      class="list-group-item d-flex justify-content-between align-items-start"
    >
      
  
      <a
        class="fw-bold font-weight-bold"
        data-id="2"
        href="http://example.com/test/1610699100"
        rel="noopener noreferrer"
        target="_blank"
      >
        
    Lorem ipsum 2021-01-15T08:25:00Z
  
      </a>
      
  
      <button
        class="btn btn-primary btn-sm"
        data-bs-target="#modal"
        data-bs-toggle="modal"
        data-id="http://example.com/test/1610699100"
        type="button"
      >
        
    Preview
  
      </button>
      

    </li>
    


    <li
      class="list-group-item d-flex justify-content-between align-items-start"
    >
      
  
      <a
        class="fw-bold font-weight-bold"
        data-id="2"
        href="http://example.com/test/1610699040"
        rel="noopener noreferrer"
        target="_blank"
      >
        
    Lorem ipsum 2021-01-15T08:24:00Z
  
      </a>
      
  
      <button
        class="btn btn-primary btn-sm"
        data-bs-target="#modal"
        data-bs-toggle="modal"
        data-id="http://example.com/test/1610699040"
        type="button"
      >
        
    Preview
  
      </button>
      

    </li>
    

    
  </ul>
  
  
</div>
`;

exports[`valid rss urls 1`] = `
<div
  class="col-md-10 col-lg-8 mx-auto feeds"
  data-component="feeds"
  data-testid="feeds"
>
  
    
  <h2>
    Feeds
  </h2>
  
    
  <ul
    class="list-group mb-5"
  >
    
  
          
    <li
      class="list-group-item"
    >
      
            
      <h3>
        Lorem ipsum feed for an interval of 1 minutes with 10 item(s)
      </h3>
      
            
      <p>
        This is a constantly updating lorem ipsum feed
      </p>
      
          
    </li>
    
      
    
  </ul>
  
  
</div>
`;

exports[`valid rss urls 2`] = `
<div
  class="col-md-10 col-lg-8 mx-auto posts"
  data-component="posts"
  data-testid="posts"
>
  
    
  <h2>
    Posts
  </h2>
  
    
  <ul
    class="list-group"
  >
    
      

    <li
      class="list-group-item d-flex justify-content-between align-items-start"
    >
      
  
      <a
        class="fw-bold font-weight-bold"
        data-id="2"
        href="http://example.com/test/1610699580"
        rel="noopener noreferrer"
        target="_blank"
      >
        
    Lorem ipsum 2021-01-15T08:33:00Z
  
      </a>
      
  
      <button
        class="btn btn-primary btn-sm"
        data-bs-target="#modal"
        data-bs-toggle="modal"
        data-id="http://example.com/test/1610699580"
        type="button"
      >
        
    Preview
  
      </button>
      

    </li>
    


    <li
      class="list-group-item d-flex justify-content-between align-items-start"
    >
      
  
      <a
        class="fw-bold font-weight-bold"
        data-id="2"
        href="http://example.com/test/1610699520"
        rel="noopener noreferrer"
        target="_blank"
      >
        
    Lorem ipsum 2021-01-15T08:32:00Z
  
      </a>
      
  
      <button
        class="btn btn-primary btn-sm"
        data-bs-target="#modal"
        data-bs-toggle="modal"
        data-id="http://example.com/test/1610699520"
        type="button"
      >
        
    Preview
  
      </button>
      

    </li>
    


    <li
      class="list-group-item d-flex justify-content-between align-items-start"
    >
      
  
      <a
        class="fw-bold font-weight-bold"
        data-id="2"
        href="http://example.com/test/1610699460"
        rel="noopener noreferrer"
        target="_blank"
      >
        
    Lorem ipsum 2021-01-15T08:31:00Z
  
      </a>
      
  
      <button
        class="btn btn-primary btn-sm"
        data-bs-target="#modal"
        data-bs-toggle="modal"
        data-id="http://example.com/test/1610699460"
        type="button"
      >
        
    Preview
  
      </button>
      

    </li>
    


    <li
      class="list-group-item d-flex justify-content-between align-items-start"
    >
      
  
      <a
        class="fw-bold font-weight-bold"
        data-id="2"
        href="http://example.com/test/1610699400"
        rel="noopener noreferrer"
        target="_blank"
      >
        
    Lorem ipsum 2021-01-15T08:30:00Z
  
      </a>
      
  
      <button
        class="btn btn-primary btn-sm"
        data-bs-target="#modal"
        data-bs-toggle="modal"
        data-id="http://example.com/test/1610699400"
        type="button"
      >
        
    Preview
  
      </button>
      

    </li>
    


    <li
      class="list-group-item d-flex justify-content-between align-items-start"
    >
      
  
      <a
        class="fw-bold font-weight-bold"
        data-id="2"
        href="http://example.com/test/1610699340"
        rel="noopener noreferrer"
        target="_blank"
      >
        
    Lorem ipsum 2021-01-15T08:29:00Z
  
      </a>
      
  
      <button
        class="btn btn-primary btn-sm"
        data-bs-target="#modal"
        data-bs-toggle="modal"
        data-id="http://example.com/test/1610699340"
        type="button"
      >
        
    Preview
  
      </button>
      

    </li>
    


    <li
      class="list-group-item d-flex justify-content-between align-items-start"
    >
      
  
      <a
        class="fw-bold font-weight-bold"
        data-id="2"
        href="http://example.com/test/1610699280"
        rel="noopener noreferrer"
        target="_blank"
      >
        
    Lorem ipsum 2021-01-15T08:28:00Z
  
      </a>
      
  
      <button
        class="btn btn-primary btn-sm"
        data-bs-target="#modal"
        data-bs-toggle="modal"
        data-id="http://example.com/test/1610699280"
        type="button"
      >
        
    Preview
  
      </button>
      

    </li>
    


    <li
      class="list-group-item d-flex justify-content-between align-items-start"
    >
      
  
      <a
        class="fw-bold font-weight-bold"
        data-id="2"
        href="http://example.com/test/1610699220"
        rel="noopener noreferrer"
        target="_blank"
      >
        
    Lorem ipsum 2021-01-15T08:27:00Z
  
      </a>
      
  
      <button
        class="btn btn-primary btn-sm"
        data-bs-target="#modal"
        data-bs-toggle="modal"
        data-id="http://example.com/test/1610699220"
        type="button"
      >
        
    Preview
  
      </button>
      

    </li>
    


    <li
      class="list-group-item d-flex justify-content-between align-items-start"
    >
      
  
      <a
        class="fw-bold font-weight-bold"
        data-id="2"
        href="http://example.com/test/1610699160"
        rel="noopener noreferrer"
        target="_blank"
      >
        
    Lorem ipsum 2021-01-15T08:26:00Z
  
      </a>
      
  
      <button
        class="btn btn-primary btn-sm"
        data-bs-target="#modal"
        data-bs-toggle="modal"
        data-id="http://example.com/test/1610699160"
        type="button"
      >
        
    Preview
  
      </button>
      

    </li>
    


    <li
      class="list-group-item d-flex justify-content-between align-items-start"
    >
      
  
      <a
        class="fw-bold font-weight-bold"
        data-id="2"
        href="http://example.com/test/1610699100"
        rel="noopener noreferrer"
        target="_blank"
      >
        
    Lorem ipsum 2021-01-15T08:25:00Z
  
      </a>
      
  
      <button
        class="btn btn-primary btn-sm"
        data-bs-target="#modal"
        data-bs-toggle="modal"
        data-id="http://example.com/test/1610699100"
        type="button"
      >
        
    Preview
  
      </button>
      

    </li>
    


    <li
      class="list-group-item d-flex justify-content-between align-items-start"
    >
      
  
      <a
        class="fw-bold font-weight-bold"
        data-id="2"
        href="http://example.com/test/1610699040"
        rel="noopener noreferrer"
        target="_blank"
      >
        
    Lorem ipsum 2021-01-15T08:24:00Z
  
      </a>
      
  
      <button
        class="btn btn-primary btn-sm"
        data-bs-target="#modal"
        data-bs-toggle="modal"
        data-id="http://example.com/test/1610699040"
        type="button"
      >
        
    Preview
  
      </button>
      

    </li>
    

    
  </ul>
  
  
</div>
`;
